pkg/discovery/lifecycle.go

```diff
@@ -0,0 +1,46 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package discovery
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+// ConnectionSuccess is called when a node is connected to
+func (d *Discovery) ConnectionSuccess(ctx context.Context, address string, id storj.NodeID) {
+	err := d.cache.Put(context.Background(), id, pb.Node{
+		Address: &pb.NodeAddress{
+			Address: address,
+		},
+		Id: id,
+	})
+	if err != nil {
+		zap.S().Error("error adding node to cache", zap.Error(err))
+	}
+}
+
+// ConnectionFailure gets called when a node fails to be connected to
+func (d *Discovery) ConnectionFailure(ctx context.Context, id storj.NodeID) {
+	err := d.cache.Put(ctx, id, pb.Node{})
```

Comment:
When are these (`ConnectionFailure`/`ConnectionSuccess`) called as opposed to the observer versions of these functions?